\documentclass{article}

\begin{document}
\SweaveOpts{concordance=TRUE}

<<>>=
library("h5r")
library("pbh5")
library("data.table")
par(mfrow = c(3, 3))
setwd("/Users/alice/Desktop/projects/kinetics/passes")
cmpH5 = PacBioCmpH5("/Users/alice/Desktop/projects/kinetics/passes/chr21_P6.cmp.h5")
passesStat <-
  read.table(
    "/Users/alice/Desktop/projects/kinetics/passes/21.EmptyFeatureOnly.mf.gff.intersect"
  )

linkage <- paste(cmpH5$movieName, cmpH5$moleculeID, sep = "_")
association <-
  as.data.frame(cbind(cmpH5$ID, as.numeric(factor(linkage))))
association$movieName <- cmpH5$movieName
association$moleculeID <- cmpH5$moleculeID
association$alignedStrand <- cmpH5$alignedStrand
association$rStart <- cmpH5$rStart
association$rEnd <- cmpH5$rEnd
colnames(association) <-
  c("idx",
    "linkage",
    "movieName",
    "moleculeID",
    "strand",
    "rStart",
    "rEnd")
association <- as.data.table(association)

start <- 9829589
end <- 9829616
chromosome <- "chr21"

calculateIPDforWindow <-
  function(chromosome, start, end, numberOfPasses) {
    subreads <- getReadsInRange(cmpH5, chromosome, start, end)
    start<-as.integer(start)
    end<-as.integer(end)
    #print(paste(chromosome,start,end))
    #print(subreads)
    #print(paste("number of passes:",numberOfPasses))
    
    #print(length(subreads))
    myColors <- c("red", "blue")
    bind <- NULL
    
    molecules <-
      association[association$idx %in% subreads,] #keep only associated subreads
    
    for (mol in split(molecules, by = c("moleculeID"))) {
      m <- unique(mol$moleculeID)
      #print(paste("moleculeID: ",m))
      chunk <- mol[order(mol$rStart)] #sort subreads by rStart
      subreads_in_mol <- length(mol$idx)
      #print(subreads_in_mol)
      #View(getAlignmentBlock(cmpH5, chromosome, start, end))
      if (subreads_in_mol == numberOfPasses) {
        #print("==========")
        print(paste(chromosome, start, end))
        #print(paste("sufficient number of passes: ", length(mol$idx)))
        #print(chunk$strand)
        ideal_strandness0 <- rep(c(0, 1), length(chunk$strand) / 2)
        ideal_strandness1 <- rep(c(1, 0), length(chunk$strand) / 2)
        
        if (all.equal(chunk$strand, ideal_strandness0) == TRUE |
            all.equal(chunk$strand, ideal_strandness1) == TRUE) {
          #strands interleave as expected
          print("strands interleaved correctly ") #todo: add adjacency control for reads
          #print(chunk$strand)
          pass <- 0
          for (id in chunk[order(chunk$rStart)]$idx) {
            i<-sort(unique(id))
            pass <- pass + 1
            #print(paste("subread idx:",i,"pass",pass))
            tposIPD <- getByTemplatePosition(cmpH5, idx = i)
            tposIPD <- as.data.table(tposIPD)
            #print(dim(tposIPD))
            idx_chunk <- subset(tposIPD,tposIPD$position >= start & tposIPD$position <= end)
            #print(dim(idx_chunk))
            #print(nrow(idx_chunk))
            
            if (nrow(idx_chunk) == 0) {
              print("empty dataframe")
              #break #the whole molecule should be skipped
            } else {
              #print("non-empty dataframe")
              
              #print(head(idx_chunk, n = 1))
              #print(tail(idx_chunk, n = 1))
              
              s <- unique(sort(idx_chunk$strand))
              
              #bind<-qpcR:::cbind.na(bind,paste(idx_chunk$position,idx_chunk$elt))
              IPDbyPOS <-
                as.data.frame(cbind(idx_chunk$position, idx_chunk$elt))
              colnames(IPDbyPOS)[1] <- "position"
              
              insertionRows <-
                as.data.frame(idx_chunk[as.character(idx_chunk$ref) == "-", c("position", "read")])
              deletionRows <-
                as.data.frame(idx_chunk[as.character(idx_chunk$read) == "-", c("position", "ref")])
              mismatchRows <-
                as.data.frame(idx_chunk[as.character(idx_chunk$ref) != as.character(idx_chunk$read) &
                                          (as.character(idx_chunk$ref) != "-") &
                                          (as.character(idx_chunk$read) != "-"), c("position", "read")])
              
              #add data about insertions, deletions, mismatches
              IPDbyPOS <-
                merge(
                  IPDbyPOS,
                  mismatchRows,
                  all.x = TRUE,
                  all.y = TRUE,
                  suffixes = paste(i, pass, s)
                )
              IPDbyPOS <-
                merge(
                  IPDbyPOS,
                  deletionRows,
                  all.x = TRUE,
                  all.y = TRUE,
                  suffixes = paste(i, pass, s)
                )
              IPDbyPOS <-
                merge(
                  IPDbyPOS,
                  insertionRows,
                  all.x = TRUE,
                  all.y = TRUE,
                  suffixes = paste(i, pass, s)
                )
              colnames(IPDbyPOS) <-
                c("position", "insertions", "elt", "deletions")
              
              if (is.null(bind)) {
                bind <- IPDbyPOS
              } else {
                bind <- merge(
                  bind,
                  IPDbyPOS,
                  by = "position",
                  all.y = TRUE,
                  all.x = TRUE
                )
              }
              
              colnames(bind)[(ncol(bind) - 2):ncol(bind)] <-
                paste(colnames(bind)[(ncol(bind) - 2):ncol(bind)], i, pass, s)
              #
              
              plot(
                idx_chunk$position,
                idx_chunk$elt,
                col = myColors[s + 1],
                ylim = c(0, 60),
                ylab = "IPD",
                xlab = "position in reference",
                main = paste("molecule:", m, "| PASS:", pass, "|", "nan"),
                sub = paste(
                  "idx:",
                  i,
                  "strand:",
                  s,
                  "rStart:",
                  chunk[chunk$idx == i]$rStart,
                  "rEnd:",
                  chunk[chunk$idx == i]$rEnd
                )
              )
              #print(head(idx_chunk))
              
              
              
              filename <-
                paste0(paste(chromosome, start, end, sep = "-"),
                       "_",
                       m,
                       ".IPDs.txt")
              if (file.exists(filename))
                file.remove(filename) #remove file if it exists
              
              write.table(
                #only plot if molecule has multiple subreads
                as.matrix(bind),
                file = filename,
                col.names = TRUE,
                row.names = FALSE,
                quote = FALSE,
                sep = "\t",
                append = TRUE
              )
              
            }
          }
          bind<-NULL
        }
      }
    }
  }

calculateIPDforWindow("chr21", 10473900, 10473999, 4)
coordinates <-
  as.data.frame(unique(cbind(
    passesStat$V1, as.numeric(passesStat$V4), as.numeric(passesStat$V5)
  )))
by(coordinates, 1:nrow(coordinates), function(row)
  calculateIPDforWindow(paste0("chr", row[1]), row[2], row[3], 4))


#all_molecules<-unique(passesStat$V12)
#print(head(all_molecules))

#calculateIPDforWindow(chromosome,start,end)

@

<<PLOT DATA>>=
#par(mfrow = c(2, 1))

setwd("/Users/alice/Desktop/projects/kinetics/passes")
#pdf(file="/Users/alice/Desktop/projects/kinetics/errors_will/old/10000/molecules.pdf")
filenames <- list.files(pattern = "*IPDs.txt", full.names = FALSE)
#filenames<-filenames[grepl("Quad",filenames)]

myColors<-c("blue","red")

for (file in (filenames)) {
  d <- read.table(file, sep = "\t", header = TRUE)
  print(dim(d))
  
  strand<-as.numeric((unlist(strsplit(as.character(colnames(d)[3]),"[.]"))[4]))
  myCol<-myColors[as.numeric(strand+1)]
  
  plot(
    d[, 1],
    d[, 3],
    pch = '1',
    col = myCol,
    ylim = c(0, 60),
    ylab = "IPD",
    sub = "position on reference",
    main=basename(file),
    xlab="",
    las=2
  )
  
  for (c in seq(6, ncol(d), 3)) {
    print("---")
    print(c)
    strand<-as.numeric((unlist(strsplit(as.character(colnames(d)[c]),"[.]"))[4]))
    print(strand)
    myCol<-myColors[as.numeric(strand+1)]
    print(myCol)
    print(colnames(d)[c])
    mySymbol<-as.character((c/3))
    print(mySymbol)
    points(d[, 1], d[, c], pch = mySymbol, col=myCol)
  }
  
  #plot insertions
  for (c in seq(2, ncol(d), 3)) {
    points(d[,1],rep(10,length(d[,1])),pch=as.character(d[,c]),col="darkgreen")
  }
  
  #plot deletions
  for (c in seq(4, ncol(d), 3)) {
    points(d[,1],rep(9.5,length(d[,1])),pch=as.character(d[,c]),col="orange")
  }
  
  points(d[,1], rowMeans(d[, seq(3, ncol(d), 3)], na.rm = TRUE), pch = "*",col="gray",cex=2)
  
  #dt <- t(d)
  #colnames(dt) <- dt[1, ]
  #dt <- dt[-1, ]
  
  #boxplot(dt, las = 2)
}

#dev.off()
@

\end{document}