\documentclass{article}

\begin{document}
\SweaveOpts{concordance=TRUE}

<<abundance heatmap>>=
library(gplots) 
#Illumina
#at<-t(as.data.frame(read.table("/Users/polly/Desktop/projects/heterochromatin/abundance_heatmap/species.table.txt.with.header.txt",sep=" ",header=TRUE,row.names=1))) #THESE ARE TANDEMNESS DATA

#Nanopore
at<-as.data.frame(read.table("/Users/polly/Desktop/projects/heterochromatin/nanopore/nanopore_lengths/species.density.calculation",sep=" ",header=TRUE,row.names=1))[,1:39]

at[at == 0] <- NA
  
atsorted<-at[,order(colSums(at,na.rm=TRUE),-colSums(is.na(at)),decreasing=T)]
atm<-as.matrix(atsorted)
my_palette <- colorRampPalette(c("lightsalmon", "firebrick2"))(n = 1000)

repeat_order<-c("AATGG","ACTCC","AAAG","AATGGAGTGG","AATGGAATGGAGTGG","AATGGAGTGGAGTGG","AAGTGGAATGG","AATGGAATGTG","AAAGTGGAATGG","AACGTGGAATGG","AATGGAATGTGG","AATGGAATGGAATGTGG","AATGGAATGGAATGGAATGTGG","AAATGGACTCGAATGGAATCATC","AATCGAATGGAATGG","AAATGGAATCGAATGGAATCATC","AATCATCGAATGGAATCGAATGG","AATCATCGAATGGACTCGAATGG","AATCATCATGAATGGAATCGAATGG","AAATGGAATCGAATGGAATCATCATC","AAATGGAATCGAATGTAATCATCATC","AATCATCATCGAATGGAATCGAATGG","AAACATGTAAATATTTACACAGAATCTGTAT","AAATATCTACACCGCTATCTGTATGAACATGG","AAATATCTACACCGCCATCTGTATGAACATGG","AAACATGGAAATATCTACACAGCCATCTGTAT","AAACATGGAAATATCTACACCACCATCTGTAT","AAACATGGAAATATCTACACCACTATCTGTAT","AAACATGGAAATATCTACACCGCCATCTGTAT","AAACATGGAAATATCTACACCGCTATCTGTAT","AAACATGGAAATATCTACACCGCTATCTGTGT","AAACATGGAAATATCTACACTGCCATCTGTAT","AAACATGGAAATATCTACACTGCTATCTGTAT","AAACATGGAAATATCTACACCGCGATCTGTAT","AAACATGGAAATATCTACACCGCCATCTCTAT","AAACATGGAAATATCTACACCGCTATCTCTAT","AAATATCTACACAGCTATCTGTATGAACATGG","AAATATCTACACCACTATCTGTATGAACATGG","AAATATCTACACCGCTATCTGCATGAACATGG")

#check that loaded repeat names match those from the dataframe
original_names<-sort(as.character(colnames(atm)))
loaded_names<-sort(as.character(repeat_order))
stopifnot(original_names==loaded_names)

species_order<-c("Gorilla02","Chimpanzee03","Bonobo06","Human05","SumatranOrangutan04","BorneanOrangutan01")
stopifnot(sort(species_order)==sort(rownames(atm)))
atm<-atm[species_order,repeat_order] #REORDER REPEAT ORDER

cell_values<-t(atm) 
print(apply(cell_values,2,function(x) {sum(x,na.rm=TRUE)/100000})) #total repeat density of 39 most abundant repeats only

roundedStr <- function(arg1){
  rounded<-round(as.numeric(as.character(arg1)))/100000
  formatted<-as.numeric(format(rounded,format="f", big.mark=",", nsmall = 2, digits=0))
  #print(paste("formatted",formatted))
  small_numbers_correction<-as.character(paste(ifelse(formatted<0.01, "<0.01", formatted),""))
  #print(paste("small_numbers_correction",small_numbers_correction))
return(small_numbers_correction)
}

cell_labels=data.frame(lapply(as.data.frame(cell_values),roundedStr), stringsAsFactors=FALSE)

plot.new()
heatmap.2(t(atm),cellnote=cell_labels,notecol="white",dendrogram="none", Colv=FALSE, Rowv=FALSE, trace='none',col=my_palette,main=paste("nanopore list of ",length(repeat_order),"repeats with highest repeat density; log"))
@

<<length_profiles>>=
par(mfrow=c(2,3))
files <- list.files(path="/Users/alice/Desktop/projects/heterochromatin/figures/NoRF_lengths/subsets", pattern="*.table", full.names=T, recursive=FALSE)

querybp<-list()    
rlengths<-list()

for (f in files) {
    t <- read.table(f, header=T, sep="\t", comment.char="") # load file
    # apply function
    name<-gsub(".subset.fasta.table","",as.character(basename(f)))
    querybp[[name]] <- sort(as.vector(t$querybp))
    querybp[[paste0(name,"_read_length")]] <- sort(as.vector(t$seqLen))
    rlengths[[name]] <- sort(as.vector(t$seqLen))
    plot(t$querybp,main=name)
}

dat <- lapply(querybp, function(x) cbind(x = seq_along(x), y = x))

list.names <- names(dat)
lns <- sapply(dat, nrow)
dat <- as.data.frame(do.call("rbind", dat))
dat$group <- rep(list.names, lns)

library(ggplot2)
library(scales)
library(RColorBrewer)
library(plotly)

ggplot(dat, aes(x = y, y = x, colour = group)) +
    theme_bw() +
    labs(x = "length", y="index", title="Repeat motifs per 1,000 PacBio reads") +
    geom_point(size=3, shape=".") +
    geom_line(linetype = "solid", size=2) +
    scale_x_continuous(labels = comma) +
    theme(axis.text=element_text(size=12),axis.title=element_text(size=14,face="bold")) +
    scale_colour_brewer(palette = "Paired")
@


\end{document}